### ISSUE TITLE
Merge all "core" modules into one "druid-core" module

### ISSUE THREAD

**Issue by leventov** [apache/druid#4312]
After #3585 and #3647 some restructure of Druid modules and classes is needed. java-util should be merged with druid-common. On the other hand, there are some classes in both of those modules which are used only in a single other module. Such classes should be moved from java-util or druid-common to that module.

The scope of this issue is specifically java-util and druid-common. However similar restructures in other parts of the Druid project are also possible.

**Comment by himanshug** [comment]
Copied description from #5229 ...

There have been multiple discussions about Druid source having too many sub-projects which really doesn't provide much value and often leads to complications e.g. `server/src/main/java/io/druid/client/indexing/ClientXXXQuery.java` classes are equivalent to `XXXTask.java` classes in `indexing-service/` and needs to stay in sync .. more such classes are created in #5102 .
similar desire has been mentioned in https://github.com/metamx/java-util/pull/74#issuecomment-353473234 and https://github.com/druid-io/druid/pull/5102#discussion_r158430657 .

So, here is my initial idea....

1. merge `druid-api, druid-commons, java-utils, druid-processing` into `druid-core`
2. merge `druid-server, druid-indexing-service` into `druid-server`

we can merge things further if necessary.

**Comment by jihoonson** [comment]
@himanshug +1. sounds nice.

**Comment by leventov** [comment]
+1

**Comment by himanshug** [comment]
do others agree/disagree ?

**Comment by leventov** [comment]
@himanshug I don't think there could be any debate about this, it wasn't implemented just because nobody had time for this

**Comment by himanshug** [comment]
let us briefly talk about this one in next dev sync to resolve any concerns upfront.

**Comment by leventov** [comment]
@himanshug I extracted your comment into a separate issue #5284

**Comment by himanshug** [comment]
@leventov thanks.

I would like to note the motivation around not merging druid-server druid-indexing-service. currently, Hive depends only on druid-server to power Hive-Druid integration . druid-indexing-service depends on indexing-hadoop and subsequently hadoop, so merge would require Hive to pull-in everything which might create conflicts and also bloat the hive packaging.

**Comment by himanshug** [comment]
another motivation for merging api, processing etc inside a core module - https://github.com/druid-io/druid/pull/5298/commits/9a51d27faa873e44449728b20edd621bf7844b0d#diff-49ebe227dd316b6e3199aa4ba1ab20f0R35
